fpga_interchange/arch.h

```diff
@@ -0,0 +1,1121 @@
+    {
+        NPNR_ASSERT(pip != PipId());
+        return pip_to_net.find(pip) == pip_to_net.end() || pip_to_net.at(pip) == nullptr;
+    }
+
+    NetInfo *getBoundPipNet(PipId pip) const
+    {
+        NPNR_ASSERT(pip != PipId());
+        auto p2n = pip_to_net.find(pip);
+        return p2n == pip_to_net.end() ? nullptr : p2n->second;
+    }
+
+    WireId getConflictingPipWire(PipId pip) const
+    {
+        return getPipDstWire(pip);
+    }
+
+    NetInfo *getConflictingPipNet(PipId pip) const
+    {
+        auto p2n = pip_to_net.find(pip);
+        return p2n == pip_to_net.end() ? nullptr : p2n->second;
+    }
+
+    AllPipRange getPips() const
+    {
+        AllPipRange range;
+        range.b.cursor_tile = 0;
+        range.b.cursor_index = -1;
+        range.b.chip = chip_info;
+        ++range.b; //-1 and then ++ deals with the case of no wries in the first tile
+        range.e.cursor_tile = chip_info->width * chip_info->height;
+        range.e.cursor_index = 0;
+        range.e.chip = chip_info;
+        return range;
+    }
+
+    Loc getPipLocation(PipId pip) const
+    {
+        Loc loc;
+        getTileLoc(pip.tile, &loc);
+        loc.z = 0;
+        return loc;
+    }
+
+    IdString getPipName(PipId pip) const;
+
+    IdString getPipType(PipId pip) const;
+    std::vector<std::pair<IdString, std::string>> getPipAttrs(PipId pip) const;
+
+    uint32_t getPipChecksum(PipId pip) const { return pip.index; }
+
+    WireId getPipSrcWire(PipId pip) const
+    {
+        return canonicalWireId(chip_info, pip.tile, locInfo(pip).pip_data[pip.index].src_index);
+    }
+
+    WireId getPipDstWire(PipId pip) const
+    {
+        return canonicalWireId(chip_info, pip.tile, locInfo(pip).pip_data[pip.index].dst_index);
+    }
+
+    DelayInfo getPipDelay(PipId pip) const
+    {
+        return DelayInfo();
+    }
+
+    DownhillPipRange getPipsDownhill(WireId wire) const
+    {
+        DownhillPipRange range;
+        NPNR_ASSERT(wire != WireId());
+        TileWireRange twr = getTileWireRange(wire);
+        range.b.chip = chip_info;
+        range.b.twi = twr.b;
+        range.b.twi_end = twr.e;
+        range.b.cursor = -1;
+        ++range.b;
+        range.e.chip = chip_info;
+        range.e.twi = twr.e;
+        range.e.twi_end = twr.e;
+        range.e.cursor = 0;
+        return range;
+    }
+
+    UphillPipRange getPipsUphill(WireId wire) const
+    {
+        UphillPipRange range;
+        NPNR_ASSERT(wire != WireId());
+        TileWireRange twr = getTileWireRange(wire);
+        range.b.chip = chip_info;
+        range.b.twi = twr.b;
+        range.b.twi_end = twr.e;
+        range.b.cursor = -1;
+        ++range.b;
+        range.e.chip = chip_info;
+        range.e.twi = twr.e;
+        range.e.twi_end = twr.e;
+        range.e.cursor = 0;
+        return range;
+    }
+
+    // -------------------------------------------------
+
+    GroupId getGroupByName(IdString name) const { return GroupId(); }
```

Comment:
Although the group APIs have had little use in nextpnr so far; other than in the GUI; I think it would make sense to map sites to groups. This would give any site-aware code an easy way to iterate over the bels/wires/pips of a site.

Reply:
No disagreement, but my initial reading of groups were that they were relatively unused.  I'll add a TODO here.